File: hackerrank/problem solving/algorithms/2 - implementation/7 - divisible sum pairs/divisible_sum_pairs.rs
// https://www.hackerrank.com/challenges/divisible-sum-pairs/problem?isFullScreen=true

use num::Integer;
use text_io::read;

fn main() {
    let (n, k): (usize, usize) = (read!(), read!());
    let numbers: Vec<usize> = read_numbers(n);
    println!("{}", divisible_sum_pairs(&numbers, k));
}

fn read_numbers(n: usize) -> Vec<usize> {
    (0..n).map(|_| read!()).collect()
}

// n: length of numbers
// k: length of freq
// T: O(n + k)
// S: O(k) extra space
fn divisible_sum_pairs(numbers: &[usize], k: usize) -> usize {
    let freq: Vec<usize> = init_remainder_frequency(numbers, k);
    count_pairs_with_remainder_0(&freq)
        + count_complementary_remainder_pairs(&freq)
        + count_pairs_with_remainder_k_half(&freq)
}

fn init_remainder_frequency(numbers: &[usize], k: usize) -> Vec<usize> {
    let mut freq: Vec<usize> = vec![0; k];
    for &x in numbers.iter() {
        freq[x % k] += 1;
    }
    freq
}

fn count_pairs_with_remainder_0(freq: &[usize]) -> usize {
    pair_count(freq[0])
}

fn pair_count(n: usize) -> usize {
    n * (n - 1) / 2
}

fn count_complementary_remainder_pairs(freq: &[usize]) -> usize {
    let n: usize = freq.len();
    (1..(n + 1) / 2)
        .map(|i| freq[i] * freq[n - i])
        .sum()
}

fn count_pairs_with_remainder_k_half(freq: &[usize]) -> usize {
    let n: usize = freq.len();
    if n.is_even() {
        return pair_count(freq[n / 2]);
    }
    0
}

File: hackerrank/problem solving/algorithms/2 - implementation/7 - divisible sum pairs/divisible_sum_pairs.cpp
// https://www.hackerrank.com/challenges/divisible-sum-pairs/problem?isFullScreen=true
// C++23

import std;
using namespace std;

vector<int> read_numbers(int n);
int divisible_sum_pairs(const vector<int>& numbers, int k);
vector<int> init_remainder_frequency(const vector<int>& numbers, int k);
int count_pairs_with_remainder_0(const vector<int>& freq);
int pair_count(int n);
int count_complementary_remainder_pairs(const vector<int>& freq);
int count_pairs_with_remainder_k_half(const vector<int>& freq);

int main()
{
    int n, k;
    cin >> n >> k;
    vector numbers{ read_numbers(n) };
    cout << divisible_sum_pairs(numbers, k);
    return 0;
}

vector<int> read_numbers(int n)
{
    vector<int> arr(n);
    for (auto& x : arr)
        cin >> x;
    return arr;
}

// n: length of numbers
// k: length of freq
// T: O(n + k)
// S: O(k) extra space
int divisible_sum_pairs(const vector<int>& numbers, int k)
{
    vector freq{ init_remainder_frequency(numbers, k) };
    return count_pairs_with_remainder_0(freq) + count_complementary_remainder_pairs(freq) +
        count_pairs_with_remainder_k_half(freq);
}

vector<int> init_remainder_frequency(const vector<int>& numbers, int k)
{
    vector<int> freq(k);
    for (int x : numbers)
        ++freq.at(x % k);
    return freq;
}

int count_pairs_with_remainder_0(const vector<int>& freq)
{
    return pair_count(freq.front());
}

int pair_count(int n)
{
    return n * (n - 1) / 2;
}

int count_complementary_remainder_pairs(const vector<int>& freq)
{
    auto n{ freq.size() };
    return *ranges::fold_left_first(
        views::iota(1ul, (n + 1) / 2) |
        views::transform([&freq, n](auto i) { return freq.at(i) * freq.at(n - i); }),
        plus{}
    );
}

int count_pairs_with_remainder_k_half(const vector<int>& freq)
{
    auto n{ freq.size() };
    return n % 2 == 0 ? pair_count(freq.at(n / 2)) : 0;
}
